File: src/cellier/gui/anywidget/visuals/static/image_controls.js
// Unified image control ESM (self-contained; anywidget serves each _esm
// standalone).  Shared rows, a composite switch, and a page per mode.  Every
// control writes its whole dict trait back ("shared", "single" or
// "channels"); Python diffs it and emits one bus event per changed field.
//
// Rows that only mean something for some render modes, or only in 3D, hide
// themselves: rowVisible mirrors cellier.gui._image_controls.row_visible,
// with the mode lists and n_displayed_dimensions read from the model.

const THROTTLE_MS = 50;

const LABELS = {
  visible: "Visible",
  transparency_mode: "Blending",
  interpolation: "Interpolation",
  attenuation: "Attenuation",
  color_map: "Colormap",
  clim: "Contrast",
  opacity: "Opacity",
  render_mode: "Render mode",
  iso_threshold: "Threshold",
};

const MODE_FIELDS = ["color_map", "clim", "opacity", "render_mode", "iso_threshold"];

// Decimal places for the fraction-like fields (opacity, attenuation); mirrors
// cellier.gui._image_controls.FRACTION_DECIMALS.  Data-unit fields (contrast
// limits, threshold) use the model's "decimals" trait instead.
const FRACTION_DECIMALS = 2;

// The shortest the contrast and threshold tracks may be, in pixels; mirrors
// cellier.gui._image_controls.MIN_TRACK_WIDTH_PX.  The dock grows to keep it.
const MIN_TRACK_WIDTH_PX = 120;

// Fields whose rows show only in 3D; mirrors
// cellier.gui._image_controls.THREE_D_FIELDS.
const THREE_D_FIELDS = ["render_mode", "iso_threshold", "attenuation"];

function rowVisible(field, modes, nDisplayed, thresholdModes, attenuationModes) {
  if (!THREE_D_FIELDS.includes(field)) return true;
  if (nDisplayed !== 3) return false;
  if (field === "iso_threshold") return modes.some((m) => thresholdModes.includes(m));
  if (field === "attenuation") return modes.some((m) => attenuationModes.includes(m));
  return true;
}

function formatNumber(value, decimals) {
  return Number(value).toFixed(decimals);
}

function throttled(fn) {
  let timer = null;
  let pending = null;
  return function (value) {
    if (timer === null) {
      fn(value);
      timer = setTimeout(function tick() {
        if (pending !== null) {
          const v = pending;
          pending = null;
          fn(v);
          timer = setTimeout(tick, THROTTLE_MS);
        } else {
          timer = null;
        }
      }, THROTTLE_MS);
    } else {
      pending = value;
    }
  };
}

function makeCheckbox(initial, onChange) {
  const el = document.createElement("input");
  el.type = "checkbox";
  el.checked = Boolean(initial);
  el.addEventListener("change", () => onChange(el.checked));
  return { el, set: (v) => { el.checked = Boolean(v); } };
}

function makeSelect(options, initial, onChange) {
  const el = document.createElement("select");
  el.className = "cellier-app-select";
  let opts = (options || []).slice();
  if (initial && !opts.includes(initial)) opts = [initial, ...opts];
  for (const name of opts) {
    const o = document.createElement("option");
    o.value = name;
    o.textContent = name;
    el.appendChild(o);
  }
  el.value = initial;
  el.addEventListener("change", () => onChange(el.value));
  return {
    el,
    set(value) {
      const v = value || "";
      if (!Array.from(el.options).some((o) => o.value === v)) {
        const o = document.createElement("option");
        o.value = v;
        o.textContent = v;
        el.prepend(o);
      }
      el.value = v;
    },
  };
}

function makeFloatSlider(min, max, initial, onChange, decimals = FRACTION_DECIMALS) {
  const el = document.createElement("div");
  el.className = "cellier-app-row-inner";
  const inp = document.createElement("input");
  inp.type = "range";
  inp.min = min;
  inp.max = max;
  inp.step = "any";
  inp.value = initial;
  const readout = document.createElement("span");
  readout.className = "cellier-app-readout";
  readout.textContent = formatNumber(initial, decimals);
  const send = throttled(onChange);
  inp.addEventListener("input", () => {
    readout.textContent = formatNumber(inp.value, decimals);
    send(parseFloat(inp.value));
  });
  inp.addEventListener("change", () => onChange(parseFloat(inp.value)));
  el.appendChild(inp);
  el.appendChild(readout);
  return {
    el,
    set(v) {
      inp.value = v;
      readout.textContent = formatNumber(v, decimals);
    },
  };
}

function makeClimSlider(range, initial, onChange, decimals) {
  // Four numbers, as the Qt control shows: the current low and high on a line
  // above the track, and the range bounds at its two ends.
  const el = document.createElement("div");
  el.className = "cellier-clim";
  const values = document.createElement("div");
  values.className = "cellier-clim-values";
  const loReadout = document.createElement("span");
  loReadout.className = "cellier-clim-readout";
  const hiReadout = document.createElement("span");
  hiReadout.className = "cellier-clim-readout";
  values.appendChild(loReadout);
  values.appendChild(hiReadout);
  const trackRow = document.createElement("div");
  trackRow.className = "cellier-clim-row";
  const minBound = document.createElement("span");
  minBound.className = "cellier-clim-bound";
  minBound.textContent = formatNumber(range[0], decimals);
  const maxBound = document.createElement("span");
  maxBound.className = "cellier-clim-bound";
  maxBound.textContent = formatNumber(range[1], decimals);
  const track = document.createElement("div");
  track.className = "cellier-clim-track";
  track.style.minWidth = `${MIN_TRACK_WIDTH_PX}px`;
  const rail = document.createElement("div");
  rail.className = "cellier-clim-rail";
  const fill = document.createElement("div");
  fill.className = "cellier-clim-fill";
  const lo = document.createElement("input");
  const hi = document.createElement("input");
  for (const [inp, cls, v] of [[lo, "cellier-clim-lo", initial[0]], [hi, "cellier-clim-hi", initial[1]]]) {
    inp.type = "range";
    inp.className = cls;
    inp.min = range[0];
    inp.max = range[1];
    inp.step = "any";
    inp.value = v;
  }
  function updateFill() {
    const mn = parseFloat(lo.min);
    const span = parseFloat(lo.max) - mn || 1;
    fill.style.left = ((parseFloat(lo.value) - mn) / span) * 100 + "%";
    fill.style.right = ((parseFloat(lo.max) - parseFloat(hi.value)) / span) * 100 + "%";
    loReadout.textContent = formatNumber(lo.value, decimals);
    hiReadout.textContent = formatNumber(hi.value, decimals);
  }
  const current = () => [parseFloat(lo.value), parseFloat(hi.value)];
  const send = throttled(onChange);
  lo.addEventListener("input", () => {
    if (parseFloat(lo.value) > parseFloat(hi.value)) lo.value = hi.value;
    updateFill();
    send(current());
  });
  hi.addEventListener("input", () => {
    if (parseFloat(hi.value) < parseFloat(lo.value)) hi.value = lo.value;
    updateFill();
    send(current());
  });
  lo.addEventListener("change", () => onChange(current()));
  hi.addEventListener("change", () => onChange(current()));
  track.appendChild(rail);
  track.appendChild(fill);
  track.appendChild(lo);
  track.appendChild(hi);
  trackRow.appendChild(minBound);
  trackRow.appendChild(track);
  trackRow.appendChild(maxBound);
  el.appendChild(values);
  el.appendChild(trackRow);
  updateFill();
  return {
    el,
    set(v) {
      lo.value = v[0];
      hi.value = v[1];
      updateFill();
    },
  };
}

function render({ model, el }) {
  el.classList.add("cellier-image-controls");
  let guard = false;
  // "page|channel|field" -> control
  let controls = {};
  // Every row that can hide: { el, page, channel, field }.
  let hideable = [];

  function row(parent, field, control) {
    const r = document.createElement("div");
    r.className = "cellier-app-row";
    r.dataset.field = field;
    const label = document.createElement("label");
    label.className = "cellier-app-label";
    label.textContent = LABELS[field] || field;
    r.appendChild(label);
    r.appendChild(control.el);
    parent.appendChild(r);
    return r;
  }

  function hideableRow(parent, page, channel, field, control) {
    const r = row(parent, field, control);
    if (THREE_D_FIELDS.includes(field)) hideable.push({ el: r, page, channel, field });
    return r;
  }

  // The shared attenuation slider, on the single page (under the render
  // mode) and once on the composite page; both set the one shared field.
  function attenuationControl() {
    const shared = model.get("shared") || {};
    const fieldsOn = model.get("fields") || [];
    if (!("attenuation" in shared) || !fieldsOn.includes("attenuation")) return null;
    return makeFloatSlider(0.0, 10.0, shared.attenuation, (v) =>
      write("shared", null, "attenuation", v),
    );
  }

  // Show or hide every row that can hide, from the current state.  Run in
  // full on every trigger so the rule lives in one place.
  function applyVisibility() {
    const n = model.get("n_displayed_dimensions");
    const thresholdModes = model.get("threshold_modes") || [];
    const attenuationModes = model.get("attenuation_modes") || [];
    const single = model.get("single") || {};
    const channels = model.get("channels") || {};
    const channelModes = Object.values(channels).map((c) => c.render_mode);
    for (const { el: r, page, channel, field } of hideable) {
      let modes;
      if (page === "composite") modes = channelModes;
      else if (page === "channel") modes = [(channels[channel] || {}).render_mode];
      else modes = [single.render_mode];
      const visible = rowVisible(field, modes, n, thresholdModes, attenuationModes);
      r.style.display = visible ? "" : "none";
    }
  }

  // Write one field back into its dict trait.
  function write(page, channel, field, value) {
    if (guard) return;
    if (page === "channel") {
      const channels = JSON.parse(JSON.stringify(model.get("channels") || {}));
      channels[channel] = { ...(channels[channel] || {}), [field]: value };
      model.set("channels", channels);
    } else {
      model.set(page, { ...(model.get(page) || {}), [field]: value });
    }
    model.save_changes();
    if (field === "render_mode") applyVisibility();
  }

  function modeControl(page, channel, field, values) {
    const fieldsOn = model.get("fields") || [];
    if (field !== "visible" && !fieldsOn.includes(field)) return null;
    const onChange = (v) => write(page, channel, field, v);
    const v = values[field];
    if (field === "visible") return makeCheckbox(v, onChange);
    if (field === "color_map") return makeSelect(model.get("colormap_names"), v, onChange);
    const decimals = model.get("decimals");
    if (field === "clim") {
      return makeClimSlider(model.get("clim_range") || [0, 1], v, onChange, decimals);
    }
    if (field === "render_mode") return makeSelect(model.get("render_modes"), v, onChange);
    if (field === "iso_threshold") {
      // A threshold is in data units, like the contrast limits.
      const [lo, hi] = model.get("clim_range") || [0, 1];
      const slider = makeFloatSlider(lo, hi, v, onChange, decimals);
      slider.el.querySelector("input").style.minWidth = `${MIN_TRACK_WIDTH_PX}px`;
      return slider;
    }
    return makeFloatSlider(0.0, 1.0, v, onChange, FRACTION_DECIMALS);
  }

  function build() {
    el.innerHTML = "";
    controls = {};
    hideable = [];
    const fieldsOn = model.get("fields") || [];

    const title = document.createElement("div");
    title.className = "cellier-group-title";
    title.textContent = model.get("title") || "Image";
    el.appendChild(title);

    const shared = model.get("shared") || {};
    const sharedBox = document.createElement("div");
    sharedBox.className = "cellier-image-shared";
    if (fieldsOn.includes("visible")) {
      const c = makeCheckbox(shared.visible, (v) => write("shared", null, "visible", v));
      controls["shared||visible"] = c;
      row(sharedBox, "visible", c);
    }
    for (const [field, options] of [
      ["transparency_mode", model.get("transparency_modes")],
      ["interpolation", model.get("interpolations")],
    ]) {
      const c = makeSelect(options, shared[field], (v) => write("shared", null, field, v));
      controls["shared||" + field] = c;
      row(sharedBox, field, c);
    }
    el.appendChild(sharedBox);

    const switchRow = document.createElement("label");
    switchRow.className = "cellier-image-switch";
    const switchBox = document.createElement("input");
    switchBox.type = "checkbox";
    switchBox.checked = Boolean(model.get("composite"));
    switchBox.addEventListener("change", () => {
      if (guard) return;
      model.set("composite", switchBox.checked);
      model.save_changes();
    });
    switchRow.appendChild(switchBox);
    switchRow.appendChild(document.createTextNode(" Composite channels"));
    switchRow.style.display = model.get("has_channel_axis") ? "" : "none";
    controls["composite||composite"] = { set: (v) => { switchBox.checked = Boolean(v); } };
    el.appendChild(switchRow);

    const singlePage = document.createElement("div");
    singlePage.className = "cellier-image-page cellier-image-single";
    const single = model.get("single") || {};
    const attenuation = [];
    for (const field of MODE_FIELDS) {
      const c = modeControl("single", null, field, single);
      if (c === null) continue;
      controls["single||" + field] = c;
      hideableRow(singlePage, "single", null, field, c);
      if (field === "render_mode") {
        const a = attenuationControl();
        if (a !== null) {
          attenuation.push(a);
          hideableRow(singlePage, "single", null, "attenuation", a);
        }
      }
    }
    el.appendChild(singlePage);

    const compositePage = document.createElement("div");
    compositePage.className = "cellier-image-page cellier-image-composite";
    const channels = model.get("channels") || {};
    const labels = model.get("channel_labels") || {};
    const keys = Object.keys(channels).sort((a, b) => parseInt(a, 10) - parseInt(b, 10));
    if (keys.length === 0) {
      const empty = document.createElement("div");
      empty.className = "cellier-image-empty";
      empty.textContent = "No channels";
      compositePage.appendChild(empty);
    }
    for (const key of keys) {
      const group = document.createElement("div");
      group.className = "cellier-channel-row";
      const head = document.createElement("div");
      head.className = "cellier-section-title";
      head.textContent = labels[key] || `Channel ${key}`;
      group.appendChild(head);
      for (const field of ["visible", ...MODE_FIELDS]) {
        const c = modeControl("channel", key, field, channels[key]);
        if (c === null) continue;
        controls["channel|" + key + "|" + field] = c;
        hideableRow(group, "channel", key, field, c);
      }
      compositePage.appendChild(group);
    }
    const compositeAttenuation = attenuationControl();
    if (compositeAttenuation !== null) {
      // One value for every channel, so one row below them all.
      attenuation.push(compositeAttenuation);
      hideableRow(compositePage, "composite", null, "attenuation", compositeAttenuation);
    }
    if (attenuation.length > 0) {
      controls["shared||attenuation"] = {
        set: (v) => {
          for (const a of attenuation) a.set(v);
        },
      };
    }
    el.appendChild(compositePage);
    applyPage();
    applyVisibility();
  }

  function applyPage() {
    const composite = Boolean(model.get("composite"));
    for (const page of el.querySelectorAll(".cellier-image-page")) {
      const isComposite = page.classList.contains("cellier-image-composite");
      page.style.display = isComposite === composite ? "" : "none";
    }
  }

  function sync(page) {
    guard = true;
    try {
      if (page === "channels") {
        const channels = model.get("channels") || {};
        for (const [key, values] of Object.entries(channels)) {
          for (const [field, v] of Object.entries(values)) {
            const c = controls["channel|" + key + "|" + field];
            if (c) c.set(v);
          }
        }
      } else {
        for (const [field, v] of Object.entries(model.get(page) || {})) {
          const c = controls[page + "||" + field];
          if (c) c.set(v);
        }
      }
    } finally {
      guard = false;
    }
  }

  let channelKeys = Object.keys(model.get("channels") || {}).sort().join(",");
  build();
  model.on("change:shared", () => sync("shared"));
  model.on("change:single", () => {
    sync("single");
    applyVisibility();
  });
  model.on("change:channels", () => {
    const keys = Object.keys(model.get("channels") || {}).sort().join(",");
    if (keys !== channelKeys) {
      channelKeys = keys;
      build();
    } else {
      sync("channels");
      applyVisibility();
    }
  });
  for (const name of ["n_displayed_dimensions", "threshold_modes", "attenuation_modes"]) {
    model.on(`change:${name}`, applyVisibility);
  }
  model.on("change:composite", () => {
    guard = true;
    try {
      controls["composite||composite"].set(model.get("composite"));
    } finally {
      guard = false;
    }
    applyPage();
  });
  model.on("change:fields", build);
  model.on("change:title", build);
}

export default { render };

File: src/cellier/gui/_image_controls.py
"""Toolkit-neutral half of the unified image control (unified image design 3.10).

Both front ends draw the same control -- a shared section, a mode switch, a
single-mode page and a composite page -- and have to answer the same questions
about it: what it starts with, which bus event a change to each field rides
on, and how a model value is shown.  Those answers live here once.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

from cellier.events import (
    AppearanceChangedEvent,
    AppearanceUpdateEvent,
    ChannelAppearanceChangedEvent,
    ChannelAppearanceUpdateEvent,
    ImageCompositeChangedEvent,
    SingleAppearanceChangedEvent,
    SingleAppearanceUpdateEvent,
    VisualVisibilityChangedEvent,
)
from cellier.gui._colormap_util import colormap_to_str

if TYPE_CHECKING:
    from collections.abc import Iterable
    from uuid import UUID

#: Shared-section fields, in display order.
SHARED_FIELDS: tuple[str, ...] = (
    "visible",
    "transparency_mode",
    "interpolation",
    "attenuation",
)

#: Fields each mode page shows, in display order.  A channel row adds
#: ``visible`` in front.
MODE_FIELDS: tuple[str, ...] = (
    "color_map",
    "clim",
    "opacity",
    "render_mode",
    "iso_threshold",
)

#: The shared transparency modes, with ``"auto"`` standing for ``None``.
TRANSPARENCY_CHOICES: tuple[str, ...] = (
    "auto",
    "blend",
    "add",
    "multiply",
    "weighted_blend",
    "weighted_solid",
)

INTERPOLATION_CHOICES: tuple[str, ...] = ("nearest", "linear")

#: Decimal places for the fraction-like fields (opacity, attenuation), on both
#: front ends.  Fixed, unlike the data-unit fields, which follow the config's
#: ``decimals``.
FRACTION_DECIMALS: int = 2

#: The shortest the contrast-limits and iso-threshold slider tracks may be, in
#: logical pixels, on both front ends.  Their number labels take most of a
#: narrow dock's width; the dock grows to keep this much track instead.
MIN_TRACK_WIDTH_PX: int = 120

#: Render modes that use the iso threshold.  The threshold row is shown only
#: for these, and only in 3D.
THRESHOLD_MODES: tuple[str, ...] = ("iso", "smooth_iso")

#: Render modes that use the attenuation coefficient.  The attenuation row is
#: shown only for these, and only in 3D.
ATTENUATION_MODES: tuple[str, ...] = ("attenuated_mip",)

#: Fields whose rows are shown only when the scene displays three dimensions:
#: a 2D slice ignores them.
THREE_D_FIELDS: tuple[str, ...] = ("render_mode", "iso_threshold", "attenuation")

#: The display dimensionalities a scene can have.
DISPLAY_DIMENSIONS: tuple[int, ...] = (2, 3)

#: The inbound events the control listens to, per driven visual.
INBOUND_EVENT_TYPES: tuple[type, ...] = (
    AppearanceChangedEvent,
    VisualVisibilityChangedEvent,
    SingleAppearanceChangedEvent,
    ChannelAppearanceChangedEvent,
    ImageCompositeChangedEvent,
)

FIELD_LABELS: dict[str, str] = {
    "visible": "Visible",
    "transparency_mode": "Blending",
    "interpolation": "Interpolation",
    "attenuation": "Attenuation",
    "color_map": "Colormap",
    "clim": "Contrast",
    "opacity": "Opacity",
    "render_mode": "Render mode",
    "iso_threshold": "Threshold",
}

DEFAULT_COLORMAP_NAMES: list[str] = [
    "gray",
    "viridis",
    "plasma",
    "magma",
    "inferno",
    "cividis",
    "red",
    "green",
    "blue",
    "magenta",
    "cyan",
]


def validate_n_displayed_dimensions(value: Any) -> int:
    """Return *value* if it is a display dimensionality (2 or 3).

    Raises
    ------
    TypeError
        If *value* is not an int (``bool`` included).
    ValueError
        If *value* is not 2 or 3.
    """
    if isinstance(value, bool) or not isinstance(value, int):
        raise TypeError(
            f"n_displayed_dimensions must be an int, got {type(value).__name__}."
        )
    if value not in DISPLAY_DIMENSIONS:
        raise ValueError(f"n_displayed_dimensions must be 2 or 3, got {value}.")
    return value


def row_visible(
    field: str, render_modes: Iterable[str], n_displayed_dimensions: int
) -> bool:
    """Whether a row of the image control is shown.

    Both front ends call this for every row that can hide, each time anything
    it depends on changes, rather than adjusting rows one trigger at a time.

    Parameters
    ----------
    field : str
        The row's field.
    render_modes : iterable of str
        The render modes the row serves: its page's mode, or every channel's
        for the composite page's shared attenuation row.
    n_displayed_dimensions : int
        How many dimensions the scene displays (2 or 3).

    Returns
    -------
    bool
        ``render_mode`` rows are shown in 3D; ``iso_threshold`` rows in 3D for
        a mode in :data:`THRESHOLD_MODES`; ``attenuation`` rows in 3D for a
        mode in :data:`ATTENUATION_MODES`.  Every other row is always shown.
    """
    if field not in THREE_D_FIELDS:
        return True
    if n_displayed_dimensions != 3:
        return False
    if field == "iso_threshold":
        return any(mode in THRESHOLD_MODES for mode in render_modes)
    if field == "attenuation":
        return any(mode in ATTENUATION_MODES for mode in render_modes)
    return True


def displayed_dimensions(dims_state: Any) -> int | None:
    """How many dimensions *dims_state* displays; ``None`` if it does not say.

    A selection with no ``displayed_axes`` (the plane-selection stub) gives
    ``None``, and the control keeps what it had.
    """
    axes = getattr(getattr(dims_state, "selection", None), "displayed_axes", None)
    return None if axes is None else len(axes)


def display_seed(
    controller: Any, visual_ids: Iterable[UUID]
) -> tuple[tuple[UUID, ...], int]:
    """The scenes an image control follows, and what they display now.

    ``DimsChangedEvent`` fires only on a change and ``connect_widget`` does
    not replay state, so a control is seeded with the current value at
    construction and then follows the events.

    Parameters
    ----------
    controller : CellierController or None
        The controller the control is wired to.  ``None`` gives no scenes
        and 3 (every row that can hide is shown).
    visual_ids : iterable of UUID
        The visuals the control drives.

    Returns
    -------
    tuple
        ``(scene_ids, n_displayed_dimensions)``: each driven visual's scene,
        in order and without repeats, and the first scene's display
        dimensionality.  A control assumes its visuals share one.
    """
    if controller is None:
        return (), 3
    scene_ids: list[UUID] = []
    for visual_id in visual_ids:
        try:
            scene_id = controller.get_visual_scene_id(visual_id)
        except KeyError:
            continue
        if scene_id not in scene_ids:
            scene_ids.append(scene_id)
    n = None
    if scene_ids:
        n = displayed_dimensions(controller.get_scene(scene_ids[0]).dims)
    return tuple(scene_ids), 3 if n is None else n


def to_widget_value(field: str, value: Any) -> Any:
    """Show a model value in a control's terms."""
    if field == "color_map":
        return colormap_to_str(value)
    if field == "clim":
        return [float(value[0]), float(value[1])]
    if field == "transparency_mode":
        return "auto" if value is None else str(value)
    if field in ("opacity", "iso_threshold", "attenuation"):
        return float(value)
    return value


def to_model_value(field: str, value: Any) -> Any:
    """Turn a control's value back into what the model field takes."""
    if field == "clim":
        return (float(value[0]), float(value[1]))
    if field == "transparency_mode":
        return None if value in (None, "auto") else str(value)
    return value


def image_update_event(
    source_id: UUID,
    visual_id: UUID,
    page: str,
    field: str,
    value: Any,
    channel: int | None = None,
):
    """The update event that carries one edit of the image control.

    Parameters
    ----------
    source_id, visual_id : UUID
        The control's id and the visual being written.
    page : {"shared", "single", "channel"}
        Which model the field is on.
    field : str
        The field name.
    value : Any
        The control's value; converted with :func:`to_model_value`.
    channel : int or None
        The channel index, for ``page="channel"``.

    Returns
    -------
    NamedTuple
        An ``AppearanceUpdateEvent``, ``SingleAppearanceUpdateEvent`` or
        ``ChannelAppearanceUpdateEvent``.
    """
    model_value = to_model_value(field, value)
    if page == "shared":
        return AppearanceUpdateEvent(source_id, visual_id, field, model_value)
    if page == "single":
        return SingleAppearanceUpdateEvent(source_id, visual_id, field, model_value)
    if page == "channel":
        return ChannelAppearanceUpdateEvent(
            source_id, visual_id, int(channel), field, model_value
        )
    raise ValueError(f"Unknown image control page {page!r}.")


def mode_values(appearance: Any) -> dict[str, Any]:
    """A single or channel appearance, in control terms."""
    values = {
        field: to_widget_value(field, getattr(appearance, field))
        for field in MODE_FIELDS
    }
    if hasattr(appearance, "visible"):
        values["visible"] = bool(appearance.visible)
    return values


def image_control_values(
    visual: Any,
    *,
    fields: list[str] | tuple[str, ...],
    colormap_names: list[str] | None = None,
    clim_range: tuple[float, float] | None = None,
    channel_labels: dict[int, str] | None = None,
    decimals: int = 2,
) -> dict[str, Any]:
    """Everything either front end needs to build the image control.

    Parameters
    ----------
    visual : BaseImageVisual
        The representative visual.
    fields : sequence of str
        The appearance fields the config asked for.  The shared section's
        blending and interpolation rows and the mode switch are always shown.
    colormap_names : list[str] or None
        Colormap choices.  A curated default when ``None``.
    clim_range : tuple[float, float] or None
        Contrast slider bounds.  Inferred from the current limits when
        ``None``.
    channel_labels : dict[int, str] or None
        Per-channel names; ``"Channel {i}"`` when absent.
    decimals : int
        Decimal places for values in data units (the contrast limits and the
        iso threshold).  Fractions such as opacity always show 2.

    Returns
    -------
    dict
        Plain data, JSON-compatible apart from integer channel keys and
        ``"colormaps"``, which holds the models' own ``color_map`` values.
        A colormap built inline (``Colormap([...], name="white_green")``) is
        not in cmap's catalogue, so its name cannot be resolved back into a
        colormap; a toolkit that needs the real thing reads it from there,
        while the page dicts keep the JSON-safe name.
    """
    from cellier.gui._appearance_fields import literal_choices

    appearance = visual.appearance
    multiscale = hasattr(appearance, "lod_bias")
    shared = {
        "visible": bool(appearance.visible),
        "transparency_mode": to_widget_value(
            "transparency_mode", appearance.transparency_mode
        ),
        "interpolation": appearance.interpolation,
    }
    if multiscale:
        shared["attenuation"] = float(appearance.attenuation)

    channels = {int(k): mode_values(v) for k, v in visual.channels.items()}
    if clim_range is None:
        limits = [visual.single.clim, *(c.clim for c in visual.channels.values())]
        clim_range = (
            min([0.0, *(float(c[0]) for c in limits)]),
            max([1.0, *(float(c[1]) for c in limits)]),
        )
    labels = dict(channel_labels or {})
    return {
        "storage": "multiscale" if multiscale else "memory",
        "colormaps": {
            "single": getattr(visual.single, "color_map", None),
            "channels": {
                int(k): getattr(v, "color_map", None)
                for k, v in visual.channels.items()
            },
        },
        "has_channel_axis": visual.channel_axis is not None,
        "composite": bool(visual.composite),
        "fields": [str(field) for field in fields],
        "shared": shared,
        "single": mode_values(visual.single),
        "channels": channels,
        "channel_labels": {k: labels.get(k, f"Channel {k}") for k in channels},
        "render_modes": list(literal_choices(visual.single, "render_mode")),
        "transparency_modes": list(TRANSPARENCY_CHOICES),
        "interpolations": list(INTERPOLATION_CHOICES),
        "colormap_names": list(colormap_names)
        if colormap_names is not None
        else list(DEFAULT_COLORMAP_NAMES),
        "clim_range": [float(clim_range[0]), float(clim_range[1])],
        "decimals": int(decimals),
        "threshold_modes": list(THRESHOLD_MODES),
        "attenuation_modes": list(ATTENUATION_MODES),
    }


def inbound_target(event: Any) -> tuple[str, int | None, str, Any] | None:
    """Where an inbound event lands on the control.

    Returns
    -------
    tuple or None
        ``(page, channel, field, widget value)``, with ``page`` one of
        ``"shared"``, ``"single"``, ``"channel"`` or ``"composite"``; ``None``
        for an event the control does not show.  A replaced ``single`` model
        (``field_name=None``) is reported as ``("single", None, None, model)``.
    """
    if isinstance(event, VisualVisibilityChangedEvent):
        return ("shared", None, "visible", bool(event.visible))
    if isinstance(event, ImageCompositeChangedEvent):
        return ("composite", None, "composite", bool(event.composite))
    if isinstance(event, SingleAppearanceChangedEvent):
        if event.field_name is None:
            return ("single", None, None, event.new_value)
        if event.field_name not in MODE_FIELDS:
            return None
        return (
            "single",
            None,
            event.field_name,
            to_widget_value(event.field_name, event.new_value),
        )
    if isinstance(event, ChannelAppearanceChangedEvent):
        if event.field_name not in (*MODE_FIELDS, "visible"):
            return None
        return (
            "channel",
            int(event.channel_index),
            event.field_name,
            to_widget_value(event.field_name, event.new_value),
        )
    if isinstance(event, AppearanceChangedEvent):
        if event.field_name not in SHARED_FIELDS:
            return None
        return (
            "shared",
            None,
            event.field_name,
            to_widget_value(event.field_name, event.new_value),
        )
    return None
